### ISSUE TITLE
Add multithreading to OMake on systems that support it

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#85]
One of the interesting things that seems to be a BIG reason that current tests take forever (30+ minutes) is that there is a lack of multithreading support.

One of the big uses of multithreading is compiling the needed resources together, so let's say that you have a makefile ```make
all: a.exe

main:
    
main.o: $(other_sources)
```

Would compile $(other_sources) independently before compiling main.o, then use main.o and it's other sources in the rule "main" to compile them together when using GNU make, this could happen in parallel allowing for multiple builds to take place at once, therefore greatly speeding up the build process on multi-core machines. You can use this to great effect and see it's effects by attempting builds in parallel in VS2017 rather than using OMake for OrangeC. I think this would be a great inclusion for speeding up builds using OMake.

**Comment by LADSoft** [comment]
I'm making progress on this; I have a mostly working prototype that does things in parallel and you are right it is much faster.   But It is still slightly buggy and I also have to figure out why the make files don't work in parallel mode...

**Comment by LADSoft** [comment]
I've commited a version that seems functionally correct... and makefiles that work with it (assuming you don't overload the processor), but haven't upgraded the appveyor build yet.   That is for another day...

**Comment by GitMensch** [comment]
First tests with PDCurses show nice results. Will try with a bigger project which can count as "real" test once the missing parts are in place.

**Comment by chuggafan** [comment]
Testing self-build of the system with the option /j:3 gets me issues, specifically relating to failing the creation of object output directories before attempting to build files INTO those directories, seems to be an issue on my machine (note: I didn't overload my computer at all, it just seems to do it at random points depending on how much you have previously had built, starting with a clean slate gets you this issue faster if you also have more cores being used)

**Comment by LADSoft** [comment]
i actually fought with that for a long time, I thought I had licked it but that was with -j:7.   -j:3 didn't work on mine either...   I think I finally figured out how to make it deterministic though.   I'm also building in parallel on appveyor as we speak, it is about halfway through the MINGW64 target, we'll see how it goes (fingers crossed)

**Comment by chuggafan** [comment]
Here's hoping! I would imagine that it'd also be better to not use window's underlying semaphores and similar object-types to allow for more portability of the main build tools, as it stands the os functions can only use windows, but c++11 allows for you to use threads and do it in a more cross-platform way, as the build currently stands for instance, you would not be able to build for Linux, but it seems to work well and looking into semaphores was interesting in of its self. Hopefully this all works great!

**Comment by chuggafan** [comment]
The generated programs being held open problem is actually something I've encountered, an easy way to test this is attempting to Ctrl+C terminate during a build, it doesn't always happen, but it causes the issue reliably enough to make it a problem, as for OS.CPP, I totally agree that there will always be a reason to have it, but you could also drop normal syncronization with STDIO due to http://www.cplusplus.com/reference/ios/ios_base/sync_with_stdio/ and using Critical Sections may be able to be replaced by a mutex to allow for one thread to use it then signal that the next method may use it instead of using Critical sections, this would also allow for more cross-platform code and minimize the amount of work needed for porting to other OSes (the only thing I can't think of being able to be changed with a c++ item would be the semaphore, but that's a Windows construct and IDK is POSIX inherently has it)

**Comment by LADSoft** [comment]
i looked at using #include <thread> but OCC doesn't support it yet.   Needs pthreads and I'll have to try to find an unencumbered implementation that compiles on win32...

**Comment by chuggafan** [comment]
? I don't exactly understand what you mean by that...

**Comment by LADSoft** [comment]
Sorry, I misstyped.   I was going to use the thread header and do a minor rewrite to omake, but I didn't implement C++ 11 threading and had kind of forgotten that.   It needs pthreads...   I don't suppose you know a pthreads implementation I could use?

**Comment by chuggafan** [comment]
Hmmm, the libcxx implementation you seem to have pulled has c++11 threading support from all the looks at it, if it doesn't you could always upgrade to a newer implementation of it and then use those headers as a supporting library. https://github.com/llvm-mirror/libcxx with https://github.com/llvm-mirror/libcxxabi/tree/master/src which falls under the BSD-esque license allowing you to import it and use it, if you really wanted to you could take it as the opportunity to support c++17, but that shouldn't be neccessary.

**Comment by LADSoft** [comment]
yeah C++11 version of libcxx I have does support threading...   the problem is it isn't a standalone threading library, it is layered on top of another library called 'pthreads'.   So I have to get a pthreads implementation to make this work...   the ones I immediately came up with had either bad licensing or relied on 'posix' functions that don't exist in the orange C rtl at this point.   The latter problem is going to be addressed in issue #114 but none of this is going to get done quickly unless I find a more amenable version of pthreads...

**Comment by chuggafan** [comment]
Hmm, pthreads is a POSIX threading system, whereas newer versions of libcxx (which can be compiled for c++11, I do believe) can run using native windows threads (if you look at https://github.com/llvm-mirror/libcxx/blob/master/src/support/win32/thread_win32.cpp it shows how they implemented it, you could backport these changes for your current version of the threading library) This all would allow you to instead of using another library (pthreads, which would be a loss in some ways as that's an extra depedency that isn't required, considering that this is a solved problem) and instead use your libcxx (with the extra support LLVM added to natively support windows)

**Comment by LADSoft** [comment]
I enabled parallel on the CI server and it seems much better.   It got through the MINGW64 part of the build with no issues...   cut 20 minutes out of the build with just j=4.  assuming there are no problems with the next half I'll be liberal with pushes over the weekend and we'll see if it is stable.

I figured out that it won't be hard to port enough of the latest libcxx C++ headers backwards to support threading...  but the win32 stuff uses fibers which my headers don't support :(.   Good news though is the C++ stuff looks earily similar to the C-11 threading, which is already supported by orangec...   so I should be able to write a wrapper around that.   I may not get to it until next week though as there are a lot of open issues ahead of it :)

**Comment by chuggafan** [comment]
Hmm, you COULD fix the fibers part not being able to be used by adding the https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/ headers found here and adding the ZPL copyright notice, as it IS compatible with GPL Licensing and AFAIK should work on all platforms supported by ORANGEC already, although the newest libcxx also would mean that building it would include support for c++17 features in the binary (because those features aren't IFDEF'd out because that'd be so much more code), meaning more DLL bloat, I think I'll work on doing some porting of header files from MINGW64 + libcxx to allow this, but my testing with just libcxx being updated has run into a lot of problems (some consisting of __declspec being used, which can be overridden, others consisting of the way they define force inlining)

**Comment by LADSoft** [comment]
the problem I have with GPL'd code with respect to the runtime library is that I'm really trying to leave it open for people to use the runtime library without open sourcing their entire project.   If I use other people's GPL'd code in the RTL I can't do that.    

As far as LIBCXX for C++2017 I tried to compile it briefly and it doesn't look too bad, I just don't want to take that on until i move to C++17.   I know it is something I eventually have to do though...   but I did a little bit of testing late the other night and discovered it may be possible to simply take the thread and mutex implementations from C++17 and backport them while leaving the rest of the libcxx implementation alone.   I had some success compiling at least the headers that way.   Although didn't get the mutex header compiling because of time constraints...   the thread headers and __mutex_base did compile though.   The C++ files wouldn't compile for lack of either PTHREADS or Windows Fibers.   

At this point I noticed two things:  one you can define a file __external_threading and have it used by #defining _LIBCPP_HAS_THREAD_API_EXTERNAL (presumably in the orangec portion of __config)  and two the support in C++, e.g. condition variables, threads, mutexes, all seems seems similarly defined in C11.   Since OrangeC supports C11 I thought well maybe the *easy* path would just be to write a new __external_threading file which patches the existing C11 functions into the C++ structure.   That has the advantage that it is immediately portable to any compiler/OS that supports C11...   so that is the path I was going to go down assuming I could get mutex.h to compile...   

if you are interested in looking at it from that angle I can make a couple of files I altered available in the repository...   the mods at this point are safe and won't affect anything else...

**Comment by chuggafan** [comment]
Nice! Better idea than what I had, I was originally going to attempt to use LIBCXX (which completely works here) and attempt to bruteforce in the compiler defines, then attack at it (at release 5 of libcxx/clang, which seems to be a good bet for when they only started experimental support of c++17 and had complete c++14 support), but looking into it more (especially now that I realize you had customized the config for OrangeC to be able to be supported), I might be able to cook something up over the next week, but I am unsure if I will succeed entirely, mostly due to being pressed for time in many areas. hopefully this all goes well. Thanks for all of your hard work, by the way.

**Comment by LADSoft** [comment]
you are welcome.   I've really had a lot of fun with this compiler package over the years but it has been a bit of effort :) 

I checked in the modified files in the above checkin...   you need copy in the threading/mutex files from the C++2017 version of libcxx over top of the existing ones, then #define NEW_ORANGEC_THREADING_BEHAVIOR.   It should at least compile 'thread' header at that point which drags in __mutex_base and __threading_support.   Haven't got mutex header working though...

**Comment by LADSoft** [comment]
after fighting with builds most of the weekend (I broke the builds in various ways while fixing other issues) and then doing a partial rewrite of omake and the make files, parallel builds seem more stable.  The CI server has done three parallel builds in a row and is working on a fourth...
